stackslib/src/net/api/postblock_proposal.rs

```diff
@@ -0,0 +1,281 @@
+use stacks_common::util::hash::{hex_bytes, Hash160, Sha256Sum};
+use stacks_common::util::retry::BoundReader;
+
+use crate::burnchains::affirmation::AffirmationMap;
+use crate::burnchains::Txid;
+use crate::chainstate::burn::db::sortdb::SortitionDB;
+use crate::chainstate::nakamoto::NakamotoBlock;
+use crate::chainstate::stacks::db::blocks::MINIMUM_TX_FEE_RATE_PER_BYTE;
+use crate::chainstate::stacks::db::StacksChainState;
+use crate::chainstate::stacks::{
+    StacksBlock, StacksBlockHeader, StacksTransaction, TransactionPayload,
+};
+use crate::core::mempool::MemPoolDB;
+use crate::cost_estimates::FeeRateEstimate;
+use crate::net::http::{
+    parse_json, Error, HttpBadRequest, HttpContentType, HttpNotFound, HttpRequest,
+    HttpRequestContents, HttpRequestPreamble, HttpResponse, HttpResponseContents,
+    HttpResponsePayload, HttpResponsePreamble, HttpServerError,
+};
+use crate::net::httpcore::{
+    request, HttpPreambleExtensions, HttpRequestContentsExtensions, RPCRequestHandler,
+    StacksHttpRequest, StacksHttpResponse,
+};
+use crate::net::p2p::PeerNetwork;
+use crate::net::relay::Relayer;
+use crate::net::{
+    Attachment, BlocksData, BlocksDatum, Error as NetError, StacksMessageType, StacksNodeState,
+};
+
+/// Represents a block proposed to the `v2/block_proposal` endpoint for validation
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct NakamotoBlockProposal {
+    /// Proposed block
+    pub block: NakamotoBlock,
+    /// Identify the stacks/burnchain fork we are on
+    pub parent_consensus_hash: ConsensusHash,
+    /// These are all the microblocks that the proposed block
+    /// will confirm.
+    pub burn_tip: BurnchainHeaderHash,
+    /// This refers to the burn block that was the current tip
+    ///  at the time this proposal was constructed. In most cases,
+    ///  if this proposal is accepted, it will be "mined" in the next
+    ///  burn block.
+    pub burn_tip_height: u32,
+    /// Mainnet, Testnet, etc.
+    pub chain_id: u32,
+}
+
+impl StacksMessageCodec for NakamotoBlockProposal {
+    fn consensus_serialize<W: Write>(&self, fd: &mut W) -> Result<(), CodecError> {
+        write_next(fd, &self.block)?;
+        write_next(fd, &self.parent_consensus_hash)?;
+        write_next(fd, &self.burn_tip)?;
+        write_next(fd, &self.burn_tip_height)?;
+        write_next(fd, &self.chain_id)
+    }
+
+    fn consensus_deserialize<R: Read>(fd: &mut R) -> Result<Self, CodecError> {
+        Ok(Self {
+            block: read_next(fd)?,
+            parent_consensus_hash: read_next(fd)?,
+            burn_tip: read_next(fd)?,
+            burn_tip_height: read_next(fd)?,
+            chain_id: read_next(fd)?,
+        })
+    }
+}
+
+/// This enum is used to supply a `reason_code` for validation
+///  rejection responses. This is serialized as an enum with string
+///  type (in jsonschema terminology).
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum ValidateRejectCode {
```

Comment:
The way it works now is that we only respond to requests from `localhost`, so we know the miner is trusted. This also means that to run a Nakamoto miner, you have to run stacks-node

Maybe later we'll add support for signing to remove this restriction

Reply:
I am more from the perspective of a caller to this endpoint could provide a block with an invalid signature. A stacks signer could be provided with a BlockProposal from an invalid miner and wish to validate it. I am writing the signer end and can add this signature validation check locally if its better there? I guess this endpoint is meant more from a "syntactically" correct BlockProposal rather than any sort of signature validation? @jcnelson ?

Reply:
It does this now in the latest commit. Basically, this endpoint will do two things with a block proposal:
- Call `validate_nakamoto_block_burnchain()` to check that the block is well-formed and a valid Nakamoto block (this is where the miner signature is validate). This function already exists in `chainstate/nakamoto/mod.rs`
- Validate the blocks transactions against the current chainstate to make sure they are valid. I still have to do this...